datalake/src/main/java/com/sequenceiq/datalake/service/resize/recovery/ResizeRecoveryService.java

```
@@ -0,0 +1,90 @@
+
+    @Inject
+    private SdxReactorFlowManager sdxReactorFlowManager;
+
+    @Inject
+    private Flow2Handler flow2Handler;
+
+    @Inject
+    private FlowChainLogService flowChainLogService;
+
+    public boolean canRecover(SdxCluster sdxCluster) {
+        FlowLog flowLog = flow2Handler.getFirstStateLogfromLatestFlow(sdxCluster.getId());
+        if (entitlementService.isDatalakeResizeRecoveryEnabled(ThreadBasedUserCrnProvider.getAccountId())) {
+            if (DatalakeResizeFlowEventChainFactory.class.getSimpleName().equals(flowChainLogService.getFlowChainType(flowLog.getFlowChainId()))) {
+                SdxStatusEntity actualStatusForSdx = sdxStatusService.getActualStatusForSdx(sdxCluster);
+                if (DatalakeStatusEnum.STOP_FAILED.equals(actualStatusForSdx.getStatus())) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public SdxRecoverableResponse validateRecovery(SdxCluster sdxCluster) {
+        SdxStatusEntity actualStatusForSdx = sdxStatusService.getActualStatusForSdx(sdxCluster);
+        if (DatalakeStatusEnum.STOP_FAILED.equals(actualStatusForSdx.getStatus())) {
+            return new SdxRecoverableResponse("Resize can be recovered from a failed stop", RecoveryStatus.RECOVERABLE);
+        }
+        return new SdxRecoverableResponse("Resize can not be recovered from this point", RecoveryStatus.NON_RECOVERABLE);
+    }
+
+    public SdxRecoveryResponse triggerRecovery(SdxCluster sdxCluster, SdxRecoveryRequest sdxRecoveryRequest) {
+        SdxStatusEntity actualStatusForSdx = sdxStatusService.getActualStatusForSdx(sdxCluster);
+        if (entitlementService.isDatalakeResizeRecoveryEnabled(ThreadBasedUserCrnProvider.getAccountId())) {
+            if (DatalakeStatusEnum.STOP_FAILED.equals(actualStatusForSdx.getStatus())) {
+                return new SdxRecoveryResponse(sdxReactorFlowManager.triggerSdxStartFlow(sdxCluster));
+            }
+        }
```

All the checks for other failure cases we are actively working. Till they are implemented you could throw an exception that recovery from this failure is not yet supported.